opentreemap/api/tests.py

```diff
@@ -618,6 +636,48 @@ def test_update_plot(self):
 
         assert_reputation(self, reputation_count + 6)
 
+    def test_update_plot_with_udfs(self):
+        self._makeUdfd('abc', 'choice', choices=['a', 'b', 'c'])
+        self._makeUdfd('def', 'choice', choices=['d', 'e', 'f'])
+        test_plot = mkPlot(self.instance, self.user)
+        test_plot.width = 1
+        test_plot.length = 2
+        test_plot.geocoded_address = 'foo'
+        test_plot.udfs['abc'] = 'a'
+        test_plot.save_with_user(self.user)
+
+        self.assertEqual(test_plot.udfs['abc'], 'a')
+
+        # Include `updated_by` because the app does send it,
+        # and the endpoint must ignore it.
+        # Likewise, include `udf:def` for the same reason.
+        updated_values = {'plot':
+                          {'geom': {'y': 0.001, 'x': 0.001, 'srid': 4326},
+                           'width': 11,
+                           'length': 22,
+                           'udf:abc': 'b',
+                           'udf:def': None,
```

Comment:
This is an example dictionary sent from the iOS app

```
{
    "plot": {
        "address_street": "158 E 67th St",
        "address_city": "Los Angeles",
        "address_zip": "90003",
        "geom": {
            "y": 33.97840416393667,
            "srid": 4326,
            "x": -118.2718002275569
        }
    },
    "tree": {
        "udf:Tree Choice": "",
        "udf:Tree Multi": []
    }
}
```

It sends `""` and `[]`  as empty values, not `None`.

This tweak makes the test pass

```diff
diff --git a/opentreemap/api/plots.py b/opentreemap/api/plots.py
index 1bef051..1322bfc 100644
--- a/opentreemap/api/plots.py
+++ b/opentreemap/api/plots.py
@@ -117,7 +117,7 @@ def update_or_create_plot(request, instance, plot_id=None):
     # Strip these out, to avoid a `KeyError` when the update tries
     # to delete them.
     format_string = 'Keys must be in the form of "model.field", not "%s"'
-    keys = [key for key in data.keys() if data[key] is None]
+    keys = [key for key in data.keys() if data[key] in (None, '')]
 
     for key in keys:
         model_type, field_name = dotted_split(
diff --git a/opentreemap/api/tests.py b/opentreemap/api/tests.py
index 0c50d64..e9c2b5e 100644
--- a/opentreemap/api/tests.py
+++ b/opentreemap/api/tests.py
@@ -639,6 +639,8 @@ class UpdatePlotAndTree(OTMTestCase):
     def test_update_plot_with_udfs(self):
         self._makeUdfd('abc', 'choice', choices=['a', 'b', 'c'])
         self._makeUdfd('def', 'choice', choices=['d', 'e', 'f'])
+        self._makeUdfd('ghi', 'multichoice', choices=['g', 'h', 'i'])
+        self._makeUdfd('jkl', 'multichoice', choices=['j', 'k', 'l'])
         test_plot = mkPlot(self.instance, self.user)
         test_plot.width = 1
         test_plot.length = 2
@@ -656,7 +658,9 @@ class UpdatePlotAndTree(OTMTestCase):
                            'width': 11,
                            'length': 22,
                            'udf:abc': 'b',
-                           'udf:def': None,
+                           'udf:def': '',
+                           'udf:ghi': ['g', 'i'],
+                           'udf:jkl': [],
                            'updated_by': self.user.pk}}
 
         response = put_json("%s/instance/%s/plots/%d" %
@@ -678,6 +682,16 @@ class UpdatePlotAndTree(OTMTestCase):
         self.assertNotIn('def', test_plot.udfs)
         self.assertEqual(response_json['plot']['udf:def'], None)
 
+        # It should have added 'ghi' to the plot udfs.
+        self.assertEqual(test_plot.udfs['ghi'], ['g', 'i'])
+        # It should have sent the new value in the response.
+        self.assertEqual(response_json['plot']['udf:ghi'], ['g', 'i'])
+
+        # It should have added 'jkl' to the plot udfs.
+        self.assertEqual(test_plot.udfs['jkl'], [])
+        # It should have sent the new value in the response.
+        self.assertEqual(response_json['plot']['udf:jkl'], [])
+
     @skip("ignore pending")
     def test_update_plot_with_pending(self):
         test_plot = mkPlot(self.instance, self.user)
```

I think we need to keep the empty list, otherwise there would be no way to clear the value of a multichoice field.

Although this makes this test of plot values pass, the app still fails when using UDFs on Tree because we have added these lines to `update_or_create_plot`

```python
if model_type == 'tree':
    model = plot.current_tree
```

- `current_tree` is a function 
- At the time this is called, the tree has not been created, so `current_tree()` is going to return None.